### ISSUE TITLE
separate network connection code from raft proper

### ISSUE THREAD

**Issue by njhurst** [bakwc/PySyncObj#4]
I'm trying to build a fuzz checker that would let us run many millions of simulated runs cheaply, I started out with lots of processes and signals (STOP/CONTINUE for network latency, KILL for dead machines).  But the network code is a lot of overhead, so I'm wondering whether it would be better to just simulate the network in a single process.

It might also pay off to use UDP given we already assume unreliable transfer?

So it might make sense to pull out a class that provides listen/accept/send/recv.

**Comment by bakwc** [comment]
> so I'm wondering whether it would be better to just simulate the network in a single process
You want to try some fake network / connetion class?
> It might also pay off to use UDP given we already assume unreliable transfer?
UDP does not guarantees packets order, so we can't just change tcp to udp - it will require some improvements. I don't see any reason to use udp.
> So it might make sense to pull out a class that provides listen/accept/send/recv.
Agree, network code isn't very nice now. I suggest following:
- make Poller a global singleton
- create class (Server) (or TcpServer), move bind from syncobj to server
- make Server and Connection accept callbacks, don't use Poller directly (Poller should be used only from Server and Connection)
You can refactor it yourself if you want. Or i'll do it on weekend.

**Comment by njhurst** [comment]
TCP also does not guarantee connection order (if I understand your code correctly, you are creating a new connection for each message?)

I have already started down the path you suggest, which is encouraging.  Not sure I'll have much time before the weekend.

**Comment by bakwc** [comment]
> you are creating a new connection for each message?

No, there is no new connection for each message - only a single connection for each node (and some reconnect stuff if there is some network issues / timeouts / etc.).

> I have already started down the path you suggest, which is encouraging. Not sure I'll have much time before the weekend.

Super, it's not a critical issue so I'll wait for your pull request, don't hurry. Btw, my skype is filippfg - you contact me there if you have any question / etc.

**Comment by JustAnotherArchivist** [comment]
I'm two years late to the discussion, but I'd like to point out that Raft itself does not require that packets are delivered in order. The end of chapter 3.3 in Diego Ongaro's thesis mentions this (emphasis mine):

> Raft assumes RPC requests and responses may be lost in the network; it is the requester’s responsibility to retry the RPC if it does not receive a response in a timely manner. Servers issue RPCs in parallel for best performance, and **Raft does not assume the network preserves ordering between RPCs**.

So if Raft is implemented according to the thesis, UDP should work just fine.

**Comment by bakwc** [comment]
Agree, may be switching to UDP is not a bad idea.

**Comment by JustAnotherArchivist** [comment]
I've created a separate issue for the potential switch to UDP: #81.

Let's get back to the original issue here: I'd also like it if the networking code was completely separated from the Raft code in PySyncObj. In addition to the fuzzing idea mentioned by @njhurst, I also want to use a different method for communcation (see below). I saw that you started working on this in b57c1144 two years ago, but looking at the code, it's still tightly coupled. Specifically, `SyncObj` is directly bound to `TcpServer` and `Node`, and it doesn't appear to be possible to implement your own networking stack except through ugly monkey-patching. Here's what I propose:

* Make `Node` an abstract base class for representing a node (either self or another server in the cluster). All networking code is implemented in subclasses of `Node`, e.g. `TCPNode` for the currently used TCP transport, which runs the TCP server and (as necessary) clients. Interaction with `SyncObj` happens as it currently does, through the `send` method for sending data to a node and the `onMessageReceived` callback for incoming data.
* Add a new argument `nodeClass` or similar to `SyncObj.__init__`, which defaults to `TCPNode` and is used instead of directly `Node` for instantiating the node object. This allows switching to a custom network stack easily by subclassing `Node` and passing that class into `SyncObj`.
* Allow passing `Node` objects into `SyncObj` instead of `selfNodeAddr` and `otherNodesAddrs`.
* Support arbitrary identifiers for nodes rather than the IP address. I haven't looked yet into what changes would be necessary for this. If all network code is moved from `SyncObj` to `Node` subclasses, this should become trivial as nodes would be identified either by the `Node` object itself or by whatever identifier (first argument?) the relevant `Node` subclass uses.

My use case for these things is using TLS for internodal communication with the nodes being identified by their certificate fingerprint rather than the IP address. But these changes would make the code flexible enough to allow for pretty much any setup I can imagine currently.

What do you think of this? I'll happily implement it myself and send a PR.

**Comment by bakwc** [comment]
Seems cool! I'll be glad to merge your PR.

**Comment by herry13** [comment]
@JustAnotherArchivist @bakwc is there any progress on this issue as well as #81?

I'm interested because, in my usecase, the services are located behind NATs. This refactoring allows me to implement UDP punchole or NAT-PMP in `TCPNode` or `UDPNode` class.

**Comment by JustAnotherArchivist** [comment]
@herry13 I've been working on this, and it's almost complete. I still need to move some more code and do a lot of testing (in particular regarding zero-downtime upgrades). I can't give you any ETA yet though.

The code actually turned out slightly different from how I described it above. `Node` is now just a container for storing information regarding a node in attributes; for the default `TCPNode`, that's just `address` (`'host:port'`), which also gets processed into `host`, `ip`, and `port` attributes. The networking code lives in a `Transport` (sub)class instead. This simplifies handling both incoming and outgoing connections.
The base `Transport` interface is very generic and doesn't really need the `SyncObj` at all. It has methods for sending a message to a `Node`, setting a callback on receiving a message, adding or dropping a `Node`, and setting a callback for when a readonly node connected. With this, it's possible to implement a transport without ever using `SyncObj` directly (only through the callback).
The `TCPTransport` still uses the `SyncObj` directly though due to the configuration settings (bind address, timeout, buffer sizes, and connection retry time), the encryptor, and the `syncobj_admin` interaction. I've considered abstracting this away as well, but it's quite specific for this transport and difficult to generalise in a way that's flexible enough. If you have any thoughts on this, please let me know. Of course, we can also refactor that again afterwards; getting the network code out of `SyncObj` is the primary goal at the moment.

I have not done anything regarding #81. But once this networking separation is done, that should be very easy to implement.

**Comment by bakwc** [comment]
I'd like to leave this method, by default transport could just do nothing and consider that it is binded from start. Some transports may require some time for initialization, it is a good feature to be able to check - if transport is ready or not.

**Comment by JustAnotherArchivist** [comment]
Alright. I'll name the `Transport` method `waitReady` then to make it more general, and it'll raise a `TransportNotReadyError` (which can then be caught in `SyncObj` and converted into a `SyncObjException` if necessary). Is it okay if I also rename the method in `SyncObj` to `waitReady` (while keeping `waitBinded` as a deprecated alias for backward compatibility)?
